src/sagemaker/workflow/pipeline_validation.py

```diff
@@ -0,0 +1,55 @@
+
+from __future__ import absolute_import
+
+from typing import Dict, Union, List
+
+import attr
+
+import botocore.loaders
+from botocore.exceptions import ClientError, ValidationError
+from typing import Any, Dict, List, Sequence, Union, Optional
+
+# from sagemaker.workflow.pipeline import Pipeline
+from sagemaker.workflow.steps import StepTypeEnum
+from sagemaker.workflow.steps import Step
+from sagemaker.workflow.step_collections import StepCollection
+
+class PipelineValidation:
+    """
+
+    Extends pipeline for client-side validation
+
+    Attributes:
+        _shapes_map dict(): dictionary of botocore model restrictions to check at compile time
+
+    """
+
+    _shapes_map = dict()
+
+    def __init__(self,
+                 steps: Sequence[Union[Step, StepCollection]] = attr.ib(factory=list)
+                 ):
+        """ Loads up the shapes from the botocore service model. """
+        loader = botocore.loaders.Loader()
+        sagemaker_model = loader.load_service_model("sagemaker", "service-2")
+        self._shapes_map = sagemaker_model["shapes"]
+        self._steps = steps
+
+    def validate_pipeline_inputs(self):
+        """ Validate pipeline definition as a series of method checks """
+        try:
+            self._validate_processing_steps()
```

Comment:
Can we move the step loop here? This way we only need to loop each step once.
Otherwise, there will be multiple loops, as we have to loop each step in `._validate_training_steps`, `.validate_tuning_steps`...